lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,318 @@
+	ignoresUpdate         chan *ignore.Matcher
+	// Keeps track the events that are tracked within a directory for event
+	// aggregation. The directory itself is not (yet) to be scanned.
+	trackedDirs map[string]FsEventsBatch
+	stop        chan struct{}
+}
+
+const (
+	fastNotifyDelay = time.Duration(500) * time.Millisecond
+	maxFiles        = 512
+	maxFilesPerDir  = 128
+)
+
+func NewFsWatcher(folderPath string, folderID string, ignores *ignore.Matcher,
+	slowNotifyDelayS int) *FsWatcher {
+	if slowNotifyDelayS == 0 {
+		slowNotifyDelayS = 60 * 60 * 24
+	}
+	return &FsWatcher{
+		folderPath:            folderPath,
+		notifyModelChan:       nil,
+		fsEvents:              make(FsEventsBatch),
+		fsEventChan:           nil,
+		WatchingFs:            false,
+		notifyDelay:           fastNotifyDelay,
+		slowNotifyDelay:       time.Duration(slowNotifyDelayS) * time.Second,
+		notifyTimerNeedsReset: false,
+		inProgress:            make(map[string]struct{}),
+		folderID:              folderID,
+		ignores:               ignores,
+		ignoresUpdate:         make(chan *ignore.Matcher),
+		trackedDirs:           make(map[string]FsEventsBatch),
+		stop:                  make(chan struct{}),
+	}
+}
+
+func (watcher *FsWatcher) StartWatchingFilesystem() (<-chan FsEventsBatch, error) {
+	fsEventChan, err := watcher.setupNotifications()
+	notifyModelChan := make(chan FsEventsBatch)
+	watcher.notifyModelChan = notifyModelChan
+	if err == nil {
+		watcher.WatchingFs = true
+		watcher.fsEventChan = fsEventChan
+		go watcher.watchFilesystem()
+	}
+	return notifyModelChan, err
+}
+
+func (watcher *FsWatcher) setupNotifications() (chan notify.EventInfo, error) {
+	c := make(chan notify.EventInfo, maxFiles)
+	absShouldIgnore := func(absPath string) bool {
+		relPath, _ := filepath.Rel(watcher.folderPath, absPath)
+		return watcher.ignores.ShouldIgnore(relPath)
+	}
+	if err := notify.WatchWithFilter(filepath.Join(watcher.folderPath, "..."),
+		c, absShouldIgnore, notify.All); err != nil {
+		notify.Stop(c)
+		close(c)
+		return nil, interpretNotifyWatchError(err, watcher.folderPath)
+	}
+	l.Infoln(watcher, "Started FsWatcher")
+	return c, nil
+}
+
+func (watcher *FsWatcher) watchFilesystem() {
+	watcher.notifyTimer = time.NewTimer(watcher.notifyDelay)
+	defer watcher.notifyTimer.Stop()
+	inProgressItemSubscription := events.Default.Subscribe(
+		events.ItemStarted | events.ItemFinished)
+	for {
+		watcher.resetNotifyTimerIfNeeded()
+		select {
+		case event, _ := <-watcher.fsEventChan:
+			watcher.newFsEvent(event.Path())
+		case <-watcher.notifyTimer.C:
+			watcher.actOnTimer()
+		case event := <-inProgressItemSubscription.C():
+			watcher.updateInProgressSet(event)
+		case ignores := <-watcher.ignoresUpdate:
+			watcher.ignores = ignores
+		case <-watcher.stop:
+			notify.Stop(watcher.fsEventChan)
+			return
+		}
+	}
+}
+
+func (watcher *FsWatcher) Stop() {
+	if watcher.WatchingFs {
+		watcher.WatchingFs = false
+		watcher.stop <- struct{}{}
```

Comment:
Currently an `FsWatcher` instance is never restarted, as folders are restarted on any configuration change. However by exposing `StartWatchingFilesystem` and `Stop` it is implied that it could be restarted (for example if in the future syncthing starts to only restart/reset the parts that are actually needed on configuration change). The `WatchingFs` bool is only used in `Stop` and `StartWatchingFilesystem`, so currently there should be no race, but it could be created by running those exported functions concurrently.

Your proposition is to remove the `WatchingFs` member of `FsWatcher` and use the `stop` channel to determine whether watching is currently active, correct?